### ISSUE TITLE
This SqlTransaction has completed; it is no longer usable

### ISSUE THREAD

**Issue by VagueGit** [Mimetis/Dotmim.Sync#1072]
Saw this exception in our logs for DMS 1.0.0-beta-1492. It's only thrown once in thousands of syncs by the customers we have moved onto this beta, so it may just be one of those things.

The full exception message is
`Dotmim.Sync.Web.Client.HttpSyncWebException: [InternalApplyThenGetChangesAsync]..[InternalApplyThenGetChangesAsync]..[InternalApplyChangesAsync]..[InternalApplyTableChangesAsync]..This SqlTransaction has completed; it is no longer usable. ----------------------- └ ─ This SqlTransaction has completed; it is no longer usable.`

The stack trace
```
Dotmim.Sync.Web.Client.WebRemoteOrchestrator <HandleSyncError>d__53.MoveNext$catch$1 ()
System.Runtime.ExceptionServices ExceptionDispatchInfo.Throw ()
System.Runtime.CompilerServices TaskAwaiter.ThrowForNonSuccess (Task)
System.Runtime.CompilerServices TaskAwaiter.ValidateEnd (Task)
DataSyncService<SyncAsync>d__19.MoveNext ()
```
I'll provide config info if that would be helpful but it looks to me like an internal DMS thing rather than a configuration thing.

**Comment by Mimetis** [comment]
Once In Thousands of syncs. That will be hard to debug :D

**Comment by symbiogenesis** [comment]
I am seeing this now with the latest preview with SQL Server on the server side, and Sqlite on the client side. I am running with .NET 8 Preview 7 on both the client and the server.

It has happened more than once, with a single user on a small number of syncs.

**Comment by symbiogenesis** [comment]
To be more clear, I have two SyncAgents, and after every sync, I wait for 1 minute, and then do another sync, in a loop.

In other words, while I sync approximately every minute, I pause the timer while it is actually executing the sync.

And on the server-side, I am using the SqlSyncChangeTrackingProvider

**Comment by Mimetis** [comment]
@symbiogenesis Are you able to reproduce it easily ?
IF yes, I would love to have a small repro and debug it.

By the way, did you test to activate the [Snapshot Isolation on SQL Server](https://learn.microsoft.com/en-us/dotnet/framework/data/adonet/sql/snapshot-isolation-in-sql-server) and see if it's resolving the issue ?

BTW, if you try the Snapshot Isolation, do not forget to use it in **DMS**, in the `SqlSyncProvider`:

``` cs
/// <summary>
/// Gets the default isolation level used during transaction
/// </summary>
public virtual IsolationLevel IsolationLevel { get; set; } = IsolationLevel.ReadCommitted;
```
And change from `IsolationLevel.ReadCommitted` to `IsolationLevel.Snapshot`

**Comment by symbiogenesis** [comment]
I do not think that I can reliably reproduce it. It is just something that I see in the logs, and my test user reported crashing.

I think your suggestion is a very good one, and I believe that may help. My only question at this point is whether Snapshot ought to be the default setting.

I will also investigate the code and see if I can make a PR to improve fault-tolerance. I know you also have some Polly-like retry logic that I will investigate.

Thanks for the help!

**Comment by symbiogenesis** [comment]
Just to be clear, and in case others need to do the same thing, I am going to try setting it on the server-side during the middleware initialization like this. Hopefully that is correct:

```csharp
public static async ValueTask<IApplicationBuilder> ConfigureDotmimSync(this WebApplication app)
{
    ArgumentNullException.ThrowIfNull(app);

    await using var scope = app.Services.CreateAsyncScope();
    await using var dbContext = scope.ServiceProvider.GetRequiredService<MyDbContext>();

    await dbContext.Database.MigrateAsync(); // Create database if needed, and run migrations if needed

    var agents = scope.ServiceProvider.GetRequiredService<IEnumerable<WebServerAgent>>();

    if (agents?.Any() != true)
    {
        throw new ArgumentException($"Cannot find any {nameof(WebServerAgent)}s");
    }

    foreach (var agent in agents)
    {
        agent.Provider.IsolationLevel = System.Data.IsolationLevel.Snapshot;
        
        // Other setup here...

    }

    return app;
}
```

**Comment by VagueGit** [comment]
I have an issue with contention on one table (unrelated to DMS). This got me to wondering why IsolationLevel.Snapshot isn't on by default. This lead me to this discussion
[Best Practice for using SNAPSHOT Isolation Level](https://learn.microsoft.com/en-us/answers/questions/1186417/best-practice-for-using-snapshot-isolation-level)
In short, Snapshot isolation has a resource overhead and can block other transactions. I think we should take care not to prioritise what we do with a database without considering the impact on others.

I chose to use IsolationLevel.ReadUncommitted as it seems to relieve contention and has the smallest footprint
[Snapshot Isolation in SQL Server](https://learn.microsoft.com/en-us/dotnet/framework/data/adonet/sql/snapshot-isolation-in-sql-server)
Full disclosure: I know very little about databases ... or anything, much

**Comment by Mimetis** [comment]
`IsolationLevel.ReadUncommitted` is really unsecure in my opinion :)

@symbiogenesis I'm reading your analysis and it makes me realize that, indeed, it's complicated to affect `CoreProvider` properties when you are using **DI** in web mode.

Maybe adding a method that inject a fully configured CoreProvider instead of using a generic would make more sense.

I mean, we still have this method:

**startup.cs:**

``` cs
// [Required]: Add a SqlSyncProvider acting as the server hub.
services.AddSyncServer<SqlSyncProvider>(connectionString, tables, options);
```

But we can imagine to add a couple of new method to have this kind of things:

``` cs
var sqlProvider = new SqlSyncProvider(connectionString);
services.AddSyncServer(sqlProvider, tables, options);
```

Would be easier to add options to your `CoreProvider`
What do you think ?

**Comment by symbiogenesis** [comment]
Interestingly, while searching around about this I found your unrelated but interesting feature of [CreateSnapshotAsync()](https://dotmimsync.readthedocs.io/Snapshot.html?highlight=snapshot) which should make client initialization faster.

I thought a simple place to run CreateSnapshotAsync() is in the setup. (That may not be the best place)

Alas, CreateSnapshotAsync doesn't work with the IsolationLevel of "Snapshot." 

It fails with the following error:

>Transaction failed because this DDL statement is not allowed inside a snapshot isolation transaction. Since metadata is not versioned, a metadata change can lead to inconsistency if mixed within snapshot isolation.

The completely default Isolation Level for Azure SQL Server is "Read Committed Snapshot."

Fixing all this actually required setting the IsolationLevel to ReadCommitted and then running CreateSnapshotAsync() and then setting the IsolationLevel to Snapshot.

Using the non-generic AddSyncServer() setup approach could make sense to allow a one-off setting, but it feels wrong to need to pass around the provider all the time, and what I have seems better than that.

However, since CreateSnapshotAsync() requires an IsolationLevel other than snapshot, it makes sense to set a specific IsolationLevel for that transaction, and then I wouldn't need to play with the IsolationLevel more than once.

**Comment by symbiogenesis** [comment]
For what it is worth, the MAUI team internally doesn't allow using the generic DI registration methods, because they use reflection and reflection is slower.

The  Microsoft.Extensions.DependencyInjection library provides a Func parameter inside of AddSingleton/AddTransient/AddScoped called implementationFactory, which allows manually constructing the object there.

You could offer something like that.

**Comment by Mimetis** [comment]
As a test, I tried to create a snapshot of [AdventureWorks](https://github.com/Mimetis/Dotmim.Sync/blob/master/CreateAdventureWorks.sql), using this code:

``` sql
ALTER DATABASE AdventureWorks SET READ_COMMITTED_SNAPSHOT ON  
```

``` cs
var snapshotDirectory = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "Snapshots");
var options = new SyncOptions();
options.BatchSize = 500;
options.SnapshotsDirectory = snapshotDirectory;
var setup = new SyncSetup(allTables);
var serverProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(serverDbName));
var remoteOrchestrator = new RemoteOrchestrator(serverProvider, options);

var snap = await remoteOrchestrator.CreateSnapshotAsync(setup);

```

And it's working fine.
Do you have an example I can use to reproduce your issue @symbiogenesis ?

**Comment by Mimetis** [comment]
> For what it is worth, I noticed that the MAUI team internally doesn't allow using the generic DI registration methods, because they use reflection and reflection is slower.

- Using `AddSyncServer<T>` is using reflection. 
- Using `AddSyncServer(CoreProvider provider)` is not using reflection

**Comment by symbiogenesis** [comment]
> As a test, I tried to create a snapshot of [AdventureWorks](https://github.com/Mimetis/Dotmim.Sync/blob/master/CreateAdventureWorks.sql), using this code:
> 
> ```sql
> ALTER DATABASE AdventureWorks SET ALLOW_SNAPSHOT_ISOLATION ON  
> ALTER DATABASE AdventureWorks SET READ_COMMITTED_SNAPSHOT ON  
> ```
> 
> ```cs
> var snapshotDirectory = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "Snapshots");
> var options = new SyncOptions();
> options.BatchSize = 500;
> options.SnapshotsDirectory = snapshotDirectory;
> var setup = new SyncSetup(allTables);
> var serverProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(serverDbName));
> var remoteOrchestrator = new RemoteOrchestrator(serverProvider, options);
> 
> var snap = await remoteOrchestrator.CreateSnapshotAsync(setup);
> ```
> 
> And it's working fine. Do you have an example I can use to reproduce your issue @symbiogenesis ?

I was using SqlSyncChangeTrackingProvider which is the main difference.

**Comment by symbiogenesis** [comment]
I will try to submit another issue with a sample project to demonstrate the Snapshot isolation level problem I saw.

And with respect to this issue of the SqlTransaction completed exception with frequent syncs, I made 3 changes on my side.

- I added a retry policy that looks for this particular exception in the isRetriable delegate
- I set the Isolation Level to Snapshot as recommended (with the caveat from above)
- I used a multi-scope approach, with a new scope that is setup by looping through all the tables in the SyncSetup and setting the SyncDirection to DownloadOnly, and then using the DownloadOnly scope for the majority of the syncs and only doing a Bidirectional sync when the app actually needs it.

With these changes, I don't see the problem anymore.
